lib/SILOptimizer/Transforms/SSADestroyHoisting.cpp

```diff
@@ -234,20 +246,131 @@ class DeinitBarriers {
+  // end_access instructions are barriers.  Do that by observing which access
+  // scopes are open when encountering a barrier.  The access scopes which are
+  // open are those for which we've seen an end_access instruction when walking
+  // backwards from the destroy_addrs.  Add these access scopes to
+  // DeinitBarriers::barrierAccessScopes.
+  //
+  // Tracking which access scopes are open consists of two parts:
+  // (1) in-block analysis
+  // (2) cross-block analysis
+  // For (1), maintain a set of access scopes which are currently open.  Insert
+  // and erase scopes when seeing begin_access and end_access instructions when
+  // they're visited in checkReachableBarrier.  A stack can't be used here
+  // because access scopes are not necessarily nested.
+  // For (2), when entering a block, the access scope is the union of all the
+  // open access scopes in the block's predecessors.
+  class FindBarrierAccessScopes {
+    DeinitBarriers &result;
+    BasicBlockSetVector destroyReachesBeginBlocks;
+    llvm::DenseMap<SILBasicBlock *, llvm::SmallPtrSet<BeginAccessInst *, 2>>
+        liveInAccessScopes;
+    llvm::SmallPtrSet<BeginAccessInst *, 2> runningLiveAccessScopes;
+
+    BackwardReachability<FindBarrierAccessScopes> reachability;
+
+  public:
+    FindBarrierAccessScopes(DeinitBarriers &result)
+        : result(result),
+          destroyReachesBeginBlocks(result.knownUses.getFunction()),
+          reachability(result.knownUses.getFunction(), *this) {
+      // Seed backward reachability with destroy points.
+      for (SILInstruction *destroy : result.knownUses.originalDestroys) {
+        reachability.initLastUse(destroy);
+      }
+    }
+
+    void markLiveAccessScopesAsBarriers() {
+      for (auto *scope : runningLiveAccessScopes) {
+        result.barrierAccessScopes.insert(scope);
+      }
+    }
+
+    bool hasReachableBegin(SILBasicBlock *block) {
+      return destroyReachesBeginBlocks.contains(block);
+    }
+
+    void markReachableBegin(SILBasicBlock *block) {
+      destroyReachesBeginBlocks.insert(block);
+      if (!runningLiveAccessScopes.empty()) {
+        liveInAccessScopes[block] = runningLiveAccessScopes;
+      }
+    }
+
+    void markReachableEnd(SILBasicBlock *block) {
+      runningLiveAccessScopes.clear();
+      for (auto *predecessor : block->getPredecessorBlocks()) {
+        auto iterator = liveInAccessScopes.find(predecessor);
+        if (iterator != liveInAccessScopes.end()) {
+          for (auto *bai : iterator->getSecond()) {
+            runningLiveAccessScopes.insert(bai);
+          }
+        }
+      }
+    }
+
+    bool checkReachableBarrier(SILInstruction *inst) {
+      if (auto *eai = dyn_cast<EndAccessInst>(inst)) {
+        runningLiveAccessScopes.insert(eai->getBeginAccess());
+      } else if (auto *bai = dyn_cast<BeginAccessInst>(inst)) {
+        runningLiveAccessScopes.erase(bai);
+      }
+      bool isBarrier = result.isBarrier(inst);
```

Comment:
There is an implicit coupling between DestroyReacability::checkReachableBarrier and FindBarrierAccessScopes::checkReachableBarrier. Although it is nicely factored, I still think it's important to comment in both places that FindBarrierAccessScopes::checkReachableBarrier must return true for every instruction in which DestroyReacability::checkReachableBarrier returns true (except the end_access instructions which are being marked as barriers). This is required for correctness.

Reply:
Thank you!  Added a comment to each describing the coupling of their implementations.